src/main/java/apoc/util/QueueUtil.java

```diff
@@ -0,0 +1,73 @@
+package apoc.util;
+
+import org.neo4j.function.ThrowingSupplier;
+
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+public class QueueUtil {
+
+    public static final int WAIT = 1;
+    public static final TimeUnit WAIT_UNIT = TimeUnit.SECONDS;
+
+    public static <T> void put(BlockingQueue<T> queue, T item, long timeoutSeconds) {
+        put(queue, item, timeoutSeconds, true, () -> {});
+    }
+
+    /**
+     * to be used instead of {@link BlockingQueue#put}
+     * @param queue
+     * @param item
+     * @param timeoutSeconds
+     * @param failWithExecption true if a {@link RuntimeException} should be thrown in case we couldn't add item into the queue within timeframe
+     * @param checkDuringOffering a callback supposed to throw an exception to terminate
+     * @param <T>
+     */
+    public static <T> void put(BlockingQueue<T> queue, T item, long timeoutSeconds, boolean failWithExecption, Runnable checkDuringOffering) {
+        withHandlingInterrupted("Queue offer interrupted before " + timeoutSeconds + " seconds", () -> {
+            long started = System.currentTimeMillis();
+            while (started + timeoutSeconds * 1000 > System.currentTimeMillis()) {
+                boolean success = queue.offer(item, WAIT, WAIT_UNIT);
+                if (success) {
+                    return null;
```

Comment:
do we have to return something from this method? or can it be just void?

Reply:
BlockingQueue.put doesn't return anything, therefore I choose void for the method signature. Since `withHandlingInterrupted` is using a generic return type, I have to return something - null in this case.

Reply:
I'm inlining that indirection anyway, see comment below.